File: app/src/test/java/activity/amigosecreto/db/DesejoModelTest.java
package activity.amigosecreto.db;

import org.junit.Test;
import static org.junit.Assert.*;

public class DesejoModelTest {

    private Desejo buildDesejo() {
        Desejo d = new Desejo();
        d.setId(1);
        d.setProduto("Livro Java");
        d.setCategoria("Livros");
        d.setLojas("Amazon");
        d.setPrecoMinimo(29.90);
        d.setPrecoMaximo(59.90);
        d.setParticipanteId(10);
        return d;
    }

    @Test
    public void defaultConstructor_allFieldsDefault() {
        Desejo d = new Desejo();
        assertEquals(0, d.getId());
        assertNull(d.getProduto());
        assertNull(d.getCategoria());
        assertNull(d.getLojas());
        assertEquals(0.0, d.getPrecoMinimo(), 0.001);
        assertEquals(0.0, d.getPrecoMaximo(), 0.001);
        assertEquals(0, d.getParticipanteId());
    }

    @Test
    public void twoArgConstructor_setsIdAndProduto() {
        Desejo d = new Desejo(5, "Fone de ouvido");
        assertEquals(5, d.getId());
        assertEquals("Fone de ouvido", d.getProduto());
    }

    @Test
    public void settersAndGetters_roundtrip() {
        Desejo d = buildDesejo();
        assertEquals(1, d.getId());
        assertEquals("Livro Java", d.getProduto());
        assertEquals("Livros", d.getCategoria());
        assertEquals("Amazon", d.getLojas());
        assertEquals(29.90, d.getPrecoMinimo(), 0.001);
        assertEquals(59.90, d.getPrecoMaximo(), 0.001);
        assertEquals(10, d.getParticipanteId());
    }

    // equals/hashCode are based on id only (plain class, not data class).
    // Identity is the DB primary key — mutable fields do not affect equality.

    @Test
    public void equals_mesmoId_isEqual() {
        Desejo a = buildDesejo();
        Desejo b = buildDesejo();
        assertEquals(a, b); // same id=1
    }

    @Test
    public void equals_differentId_notEqual() {
        Desejo a = buildDesejo();
        Desejo b = buildDesejo();
        b.setId(2);
        assertNotEquals(a, b);
    }

    @Test
    public void equals_mesmoId_diferenteProduto_aindaEqual() {
        // equals uses id only — changing other fields must not affect equality
        Desejo a = buildDesejo();
        Desejo b = buildDesejo();
        b.setProduto("Outro produto");
        assertEquals(a, b); // same id → still equal
    }

    @Test
    public void equals_mesmoId_diferenteParticipanteId_aindaEqual() {
        Desejo a = buildDesejo();
        Desejo b = buildDesejo();
        b.setParticipanteId(b.getParticipanteId() + 1);
        assertEquals(a, b); // same id → still equal
    }

    @Test
    public void equals_null_notEqual() {
        assertNotEquals(buildDesejo(), null);
    }

    @Test
    public void equals_sameInstance_isEqual() {
        Desejo d = buildDesejo();
        assertEquals(d, d);
    }

    @Test
    public void hashCode_mesmoId_sameHash() {
        Desejo a = buildDesejo();
        Desejo b = buildDesejo();
        assertEquals(a.hashCode(), b.hashCode()); // same id → same hash
    }

    @Test
    public void compareTo_smallerIdComesFirst() {
        Desejo a = new Desejo(1, "A");
        Desejo b = new Desejo(2, "B");
        assertTrue(a.compareTo(b) < 0);
        assertTrue(b.compareTo(a) > 0);
    }

    @Test
    public void compareTo_sameId_returnsZero() {
        Desejo a = new Desejo(3, "A");
        Desejo b = new Desejo(3, "B");
        assertEquals(0, a.compareTo(b));
    }

    @Test
    public void toString_containsIdAndProduto() {
        Desejo d = new Desejo(7, "Tablet");
        String s = d.toString();
        assertTrue(s.contains("id=7"));
        assertTrue(s.contains("Tablet"));
    }

    @Test
    public void nullCategoria_naoAfetaEquals() {
        // equals uses id only — null fields must not affect equality
        Desejo a = buildDesejo();
        a.setCategoria(null);
        Desejo b = buildDesejo();
        b.setCategoria(null);
        assertEquals(a, b);
    }

    @Test
    public void nullLojas_naoAfetaEquals() {
        Desejo a = buildDesejo();
        a.setLojas(null);
        Desejo b = buildDesejo();
        b.setLojas(null);
        assertEquals(a, b);
    }

    @Test
    public void nullProduto_naoAfetaEquals() {
        Desejo a = buildDesejo();
        a.setProduto(null);
        Desejo b = buildDesejo();
        b.setProduto(null);
        assertEquals(a, b);
    }

    @Test
    public void toString_nullProduto_doesNotThrow() {
        Desejo d = new Desejo(8, null);
        String s = d.toString(); // não deve lançar NullPointerException
        assertNotNull(s);
        assertTrue(s.contains("id=8"));
    }

    @Test
    public void hashCode_differentId_differentHash() {
        Desejo a = buildDesejo(); // id=1
        Desejo b = buildDesejo();
        b.setId(2);
        assertNotEquals(a.hashCode(), b.hashCode());
    }
}

File: app/src/test/java/activity/amigosecreto/db/DesejoModelTest.kt
package activity.amigosecreto.db

import org.junit.Assert.*
import org.junit.Test

class DesejoModelTest {

    private fun buildDesejo() = Desejo().apply {
        id = 1
        produto = "Livro Java"
        categoria = "Livros"
        lojas = "Amazon"
        precoMinimo = 29.90
        precoMaximo = 59.90
        participanteId = 10
    }

    @Test
    fun defaultConstructor_allFieldsDefault() {
        val d = Desejo()
        assertEquals(0, d.id)
        assertNull(d.produto)
        assertNull(d.categoria)
        assertNull(d.lojas)
        assertEquals(0.0, d.precoMinimo, 0.001)
        assertEquals(0.0, d.precoMaximo, 0.001)
        assertEquals(0, d.participanteId)
    }

    @Test
    fun twoArgConstructor_setsIdAndProduto() {
        val d = Desejo(5, "Fone de ouvido")
        assertEquals(5, d.id)
        assertEquals("Fone de ouvido", d.produto)
    }

    @Test
    fun settersAndGetters_roundtrip() {
        val d = buildDesejo()
        assertEquals(1, d.id)
        assertEquals("Livro Java", d.produto)
        assertEquals("Livros", d.categoria)
        assertEquals("Amazon", d.lojas)
        assertEquals(29.90, d.precoMinimo, 0.001)
        assertEquals(59.90, d.precoMaximo, 0.001)
        assertEquals(10, d.participanteId)
    }

    // equals/hashCode are based on id only (plain class, not data class).

    @Test
    fun equals_mesmoId_isEqual() {
        assertEquals(buildDesejo(), buildDesejo())
    }

    @Test
    fun equals_differentId_notEqual() {
        val a = buildDesejo()
        val b = buildDesejo().also { it.id = 2 }
        assertNotEquals(a, b)
    }

    @Test
    fun equals_mesmoId_diferenteProduto_aindaEqual() {
        val a = buildDesejo()
        val b = buildDesejo().also { it.produto = "Outro produto" }
        assertEquals(a, b)
    }

    @Test
    fun equals_mesmoId_diferenteParticipanteId_aindaEqual() {
        val a = buildDesejo()
        val b = buildDesejo().also { it.participanteId = it.participanteId + 1 }
        assertEquals(a, b)
    }

    @Test
    fun equals_null_notEqual() {
        assertNotEquals(buildDesejo(), null)
    }

    @Test
    fun equals_sameInstance_isEqual() {
        val d = buildDesejo()
        assertEquals(d, d)
    }

    @Test
    fun hashCode_mesmoId_sameHash() {
        assertEquals(buildDesejo().hashCode(), buildDesejo().hashCode())
    }

    @Test
    fun compareTo_smallerIdComesFirst() {
        val a = Desejo(1, "A")
        val b = Desejo(2, "B")
        assertTrue(a.compareTo(b) < 0)
        assertTrue(b.compareTo(a) > 0)
    }

    @Test
    fun compareTo_sameId_returnsZero() {
        val a = Desejo(3, "A")
        val b = Desejo(3, "B")
        assertEquals(0, a.compareTo(b))
    }

    @Test
    fun toString_containsIdAndProduto() {
        val s = Desejo(7, "Tablet").toString()
        assertTrue(s.contains("id=7"))
        assertTrue(s.contains("Tablet"))
    }

    @Test
    fun nullCategoria_naoAfetaEquals() {
        val a = buildDesejo().also { it.categoria = null }
        val b = buildDesejo().also { it.categoria = null }
        assertEquals(a, b)
    }

    @Test
    fun nullLojas_naoAfetaEquals() {
        val a = buildDesejo().also { it.lojas = null }
        val b = buildDesejo().also { it.lojas = null }
        assertEquals(a, b)
    }

    @Test
    fun nullProduto_naoAfetaEquals() {
        val a = buildDesejo().also { it.produto = null }
        val b = buildDesejo().also { it.produto = null }
        assertEquals(a, b)
    }

    @Test
    fun toString_nullProduto_doesNotThrow() {
        val s = Desejo(8, null).toString()
        assertNotNull(s)
        assertTrue(s.contains("id=8"))
    }

    @Test
    fun hashCode_differentId_differentHash() {
        val a = buildDesejo() // id=1
        val b = buildDesejo().also { it.id = 2 }
        assertNotEquals(a.hashCode(), b.hashCode())
    }
}